mycore.org/content/de/documentation/basics/basics_mcr_ocfl.md

```diff
@@ -0,0 +1,312 @@
+{{</mcr-comment>}}
+
+## Migration zu OCFL
+
+Für die Prozedur sollte sichergestellt werden, dass die Daten nicht von Nutzern bearbeitet werden.
+
+Zunächst wird der Metadaten-Manager bei dem Nativen belassen und es soll sichergestellt werden,
+dass die Repository `Main` oder die eigene Repository korrekt eingerichtet sind.
+
+Mit dem Kommando `migrate metadata to repository {Repository}` können die Daten in die angegebene Repository überspielt werden.
+Das Kommando gibt eine Statistik aus, ob alle Metadaten migriert werden konnten.
+
+Auch wenn keine Fehler zu sehen sind, sollte man trotzdem die Anzahl überprüfen, ob alle Objekte erfolgreich übertragen wurden.
+
+Wenn alles geklappt hat, dann kann man den MetadatenManager umstellen.
+
+## Migration zwischen Layouts
+
+Die Migration zwischen den OCFL Layouts funktioniert genauso wie zu OCFL,
+der einzige Unterschied ist das hierbei der Metadaten-Manager des OCFL zu nutzen ist.
+
+Um eine Repository zu Migrieren, ist die Quell-Repository mit
+`MCR.Metadata.Manager.Repository={Quell-Repository}`
+zu setzen.
+
+Die Ziel Repository muss manuell gesetzt werden (siehe [Konfiguration](#konfiguration)), diese Repository darf nicht den gleichen Typ (Repository Provider) haben wie die Quelle. Wird dies trotzdem benötigt, ist es abzuwägen, ob es nicht besser ist, die Repository einfach zu kopieren.
+
+Hierbei kann es manchmal vorkommen, dass ein `MCRUsageException` ausgelöst wird. Dies tritt dann auf,
+wenn Objekte vor der Migration zu OCFL gelöscht wurden und es versucht wird, ein Property zu lesen, da MyCoRe diese ID noch immer bekannt ist. Die Datei existiert aber nicht mehr.\
+Dies wird in zukünftigen Versionen behoben werden, <u>hat aber keinen Impakt in die Funktionsweise von OCFL</u>.
+
+## Verfügbare Repository Layouts
+### OCFL Community Extension 0003: Hashed Truncated N-tuple Trees with Object ID Encapsulating Directory for OCFL Storage Hierarchies
+
+Das *0003-hash-and-id-n-tuple-storage-layout* ist das Standard Layout in MyCoRe. Es ist ein Layout aus den offiziellen Community Extensions für OCFL. Dadurch lässt sich ein Repository mit dem *0003-hash-and-id-n-tuple-storage-layout* durch externe Programme einfach einsehen und verändern. Aus den Dateinamen wird ein Hashwert gebildet, mit dem danach die Ordnerstruktur gebildet wird.
+
+Der Vorteil ist wie erwähnt, das es kompatibel ist mit externen Tools für OCFL Repositories. Mit diesen kann man auch ohne MyCoRe Änderungen vornehmen, Dateien suchen oder ersetzen. Jedoch sollte man es vermeiden, Dateien zu löschen, da dies zu Problemen mit MyCoRe führen kann.
+
+Der Nachteil ist, dass durch die Bildung des Hashwertes ein Dateizugriff länger dauert. Durch die Nutzung des Hashwertes ist es ohne externe Tools nicht möglich, herauszufinden, in welchem Ordner sich welches Objekt befindet, sondern nur durch den root Ordner des Objektes und der `inventory.json` .
```

Comment:
_Ich hoffe, die Zugriffsdauer lässt sich vernachlässigen - würde ich nicht explizit erwähnen.
Im zweiten Teil würde ich formulieren, dass man den Root-Ordner, dessen Namen der Objekt-ID entspricht, im gesamten OCFL-Repository suchen kann  oder per Skript den Ordnernamen mit dem Hash ermitteln müsste.
Können wir den Nutzern hier ein Shell-Commando oder ähnliches zur Verfügung stellen / verlinken?_

Reply:
Ich sehe schon das es erwähnenswert ist, das es langsamer/schneller ist für Dateizugriffe und -änderungen.
mit dem Commando/Skript/Programm ist mir keines bekannt, deshalb habe ich hier keines erwähnt.

Reply:
Ich kann mir zwar nicht vorstellen, dass die zusätzliche Hash-Berechnung den Zugriff wesentlich langsamer macht. Die Zeit geht doch eigentlich für den Plattenzugriff drauf, und der ist bei beiden Layouts gleich. Aber ich will mal den "Experten" glauben, oder habt ihr schon "belastbare Zahlen".

Reply:
Na gut, dass ist schon ein messbarer Unterschied.
Ggf. macht es Sinn die Klasse HashedNTupleIdEncapsulationLayoutExtension zu überschreiben und um die Methode String mapObjectID(String objectID) einen Cache zu bauen, der die letzten 1000? Ergebnisse zwischenspeichert.
... falls die Methode mehrfach aufgerufen wird.
Das könnte interessant werden, wenn später auch Dateien in die Derivate hochgeladen werden müssen.